<a href="https://colab.research.google.com/github/eder1985/igti-bootcamp-eng-dados-cloud/blob/main/modulo_2/igti_edc_mod2_desafio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1><center>IGTI/XP Educação/Bootcamp Engenharia de Dados Cloud - Módulo 2 - Desafio
</center></h1>

### Installing Spark

Install Dependencies:


1.   Java 8
2.   Apache Spark with hadoop and
3.   Findspark (used to locate the spark in the system)


In [1]:
!apt-get install openjdk-8-jdk-headless -qq > /dev/null
!wget -q http://archive.apache.org/dist/spark/spark-3.1.1/spark-3.1.1-bin-hadoop3.2.tgz
!tar xf spark-3.1.1-bin-hadoop3.2.tgz
!pip install -q findspark

Set Environment Variables:

In [2]:
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-3.1.1-bin-hadoop3.2"

In [3]:
!ls

sample_data  spark-3.1.1-bin-hadoop3.2	spark-3.1.1-bin-hadoop3.2.tgz


In [4]:
import findspark
findspark.init()
from pyspark.sql import SparkSession
spark = SparkSession.builder.master("local[*]").getOrCreate()
spark.conf.set("spark.sql.repl.eagerEval.enabled", True) # Property used to format output tables better
spark

### Desafio

In [6]:
!wget https://homepages.dcc.ufmg.br/~pcalais/XPE/engenharia-dados/big-data-spark/desafio/dados.zip --no-check-certificate

--2023-08-10 20:14:34--  https://homepages.dcc.ufmg.br/~pcalais/XPE/engenharia-dados/big-data-spark/desafio/dados.zip
Resolving homepages.dcc.ufmg.br (homepages.dcc.ufmg.br)... 150.164.0.136
Connecting to homepages.dcc.ufmg.br (homepages.dcc.ufmg.br)|150.164.0.136|:443... connected.
  Unable to locally verify the issuer's authority.
HTTP request sent, awaiting response... 200 OK
Length: 1364337248 (1.3G) [application/zip]
Saving to: ‘dados.zip’

dados.zip           100%[===================>]   1.27G  11.4MB/s    in 2m 3s   

2023-08-10 20:16:38 (10.6 MB/s) - ‘dados.zip’ saved [1364337248/1364337248]



In [12]:
!unzip dados.zip

Archive:  dados.zip
   creating: dados/cnaes/
  inflating: dados/cnaes/cnaes.csv   
   creating: dados/estabelecimentos/
   creating: dados/estabelecimentos/estabelecimentos/
  inflating: dados/estabelecimentos/estabelecimentos/estabelecimentos-1.csv  
  inflating: dados/estabelecimentos/estabelecimentos/estabelecimentos-2.csv  
  inflating: dados/estabelecimentos/estabelecimentos/estabelecimentos-3.csv  
  inflating: dados/NOVOLAYOUTDOSDADOSABERTOSDOCNPJ.pdf  


In [13]:
!head -n 2 dados/cnaes/cnaes.csv

CNAE;DESCRICAO_CNAE
"0111301";"Cultivo de arroz"


In [21]:
df_cnaes = spark.read.csv(
    'dados/cnaes/cnaes.csv',
    header=True,
    inferSchema=True,
    sep=';',
    encoding="ISO-8859-1")

In [33]:
df_cnaes.show(5,truncate = False)

+------+---------------------------------------------------------+
|CNAE  |DESCRICAO_CNAE                                           |
+------+---------------------------------------------------------+
|111301|Cultivo de arroz                                         |
|111302|Cultivo de milho                                         |
|111303|Cultivo de trigo                                         |
|111399|Cultivo de outros cereais não especificados anteriormente|
|112101|Cultivo de algodão herbáceo                              |
+------+---------------------------------------------------------+
only showing top 5 rows



In [26]:
df_cnaes.count()

1359

In [27]:
df_cnaes.printSchema()

root
 |-- CNAE: integer (nullable = true)
 |-- DESCRICAO_CNAE: string (nullable = true)



In [28]:
!head -n 2 dados/estabelecimentos/estabelecimentos/*.csv

==> dados/estabelecimentos/estabelecimentos/estabelecimentos-1.csv <==
CNPJ_BASICO;CNPJ_ORDEM;CNPJ_DV;IDENTIFICADOR_MATRIZ_FILIAL;NOME_FANTASIA;SITUACAO_CADASTRAL;DATA_SITUACAO_CADASTRAL;MOTIVO_SITUACAO_CADASTRAL;NOME_DA_CIDADE_NO_EXTERIOR;PAIS;DATA_INICIO_ATIVIDADE;CNAE_PRINCIPAL;CNAE_SECUNDARIA;TIPO_LOGRADOURO;LOGRADOURO;NUMERO;COMPLEMENTO;BAIRRO;CEP;UF;MUNICIPIO;DDD_1;TEL_1;DDD_2;TEL_2;DDD_FAX;FAX;CORREIO_ELETRONICO;SITUACAO_ESPECIAL;DATA_SITUACAO_ESPECIAL
7396865;1;68;1;"";8;20170210;1;"";"";20050518;1412602;1411801;RUA;TUCANEIRA;30;"";DOS LAGOS;89136000;SC;8297;47;33851125;47;33851125;47.0;33851125;"";"";""

==> dados/estabelecimentos/estabelecimentos/estabelecimentos-2.csv <==
CNPJ_BASICO;CNPJ_ORDEM;CNPJ_DV;IDENTIFICADOR_MATRIZ_FILIAL;NOME_FANTASIA;SITUACAO_CADASTRAL;DATA_SITUACAO_CADASTRAL;MOTIVO_SITUACAO_CADASTRAL;NOME_DA_CIDADE_NO_EXTERIOR;PAIS;DATA_INICIO_ATIVIDADE;CNAE_PRINCIPAL;CNAE_SECUNDARIA;TIPO_LOGRADOURO;LOGRADOURO;NUMERO;COMPLEMENTO;BAIRRO;CEP;UF;MUNICIPIO;DDD_1;TEL_1

In [30]:
df_estabelecimentos = spark.read.csv(
    'dados/estabelecimentos/estabelecimentos/',
    header=True,
    inferSchema=True,
    sep=';',
    encoding="ISO-8859-1")

In [34]:
df_estabelecimentos.show(5,truncate = False)

+-----------+----------+-------+---------------------------+-------------+------------------+-----------------------+-------------------------+--------------------------+----+---------------------+--------------+---------------+---------------+-----------------+------+---------------------------------+---------+--------+---+---------+-----+--------+-----+--------+-------+--------+---------------------------+-----------------+----------------------+
|CNPJ_BASICO|CNPJ_ORDEM|CNPJ_DV|IDENTIFICADOR_MATRIZ_FILIAL|NOME_FANTASIA|SITUACAO_CADASTRAL|DATA_SITUACAO_CADASTRAL|MOTIVO_SITUACAO_CADASTRAL|NOME_DA_CIDADE_NO_EXTERIOR|PAIS|DATA_INICIO_ATIVIDADE|CNAE_PRINCIPAL|CNAE_SECUNDARIA|TIPO_LOGRADOURO|LOGRADOURO       |NUMERO|COMPLEMENTO                      |BAIRRO   |CEP     |UF |MUNICIPIO|DDD_1|TEL_1   |DDD_2|TEL_2   |DDD_FAX|FAX     |CORREIO_ELETRONICO         |SITUACAO_ESPECIAL|DATA_SITUACAO_ESPECIAL|
+-----------+----------+-------+---------------------------+-------------+------------------+-

In [32]:
df_estabelecimentos.count()

20996744

In [35]:
df_estabelecimentos.printSchema()

root
 |-- CNPJ_BASICO: integer (nullable = true)
 |-- CNPJ_ORDEM: integer (nullable = true)
 |-- CNPJ_DV: integer (nullable = true)
 |-- IDENTIFICADOR_MATRIZ_FILIAL: integer (nullable = true)
 |-- NOME_FANTASIA: string (nullable = true)
 |-- SITUACAO_CADASTRAL: integer (nullable = true)
 |-- DATA_SITUACAO_CADASTRAL: integer (nullable = true)
 |-- MOTIVO_SITUACAO_CADASTRAL: integer (nullable = true)
 |-- NOME_DA_CIDADE_NO_EXTERIOR: string (nullable = true)
 |-- PAIS: integer (nullable = true)
 |-- DATA_INICIO_ATIVIDADE: integer (nullable = true)
 |-- CNAE_PRINCIPAL: integer (nullable = true)
 |-- CNAE_SECUNDARIA: string (nullable = true)
 |-- TIPO_LOGRADOURO: string (nullable = true)
 |-- LOGRADOURO: string (nullable = true)
 |-- NUMERO: string (nullable = true)
 |-- COMPLEMENTO: string (nullable = true)
 |-- BAIRRO: string (nullable = true)
 |-- CEP: string (nullable = true)
 |-- UF: string (nullable = true)
 |-- MUNICIPIO: integer (nullable = true)
 |-- DDD_1: string (nullable = true)

#### Pergunta 1: Quantos estabelecimentos existem?
- R: 20996744


In [36]:
df_estabelecimentos.count()

20996744

#### Pergunta 2: Na tabela de estabelecimentos, quantas colunas existem e quantas são identificadas pelo spark como números?
- R: 30 e 13


In [42]:
len(df_estabelecimentos.columns)

30

In [78]:
types_list = [f.dataType for f in df_estabelecimentos.schema.fields]
numeric_types_list = [
    'ByteType',
    'ShortType',
    'IntegerType',
    'LongType',
    'FloatType',
    'DoubleType',
    'DecimalType'
    ]
filtered = filter(lambda type: str(type) in numeric_types_list, types_list)
len(list(filtered))

13

#### Pergunta 3: O formato Parquet (https://parquet.apache.org/ Links to an external site.) é um formato de armazenamento útil em engenharia de dados projetado para armazenamento e recuperação eficiente dos dados. Ele usa mecanismos de compressão para que os dados ocupem menos espaço. Usando `estabelecimentos_df.write.parquet("estabelecimentos.parquet")` , compare o tamanho do(s) arquivo(s) parquet com os arquivos CSV originais. A economia de espaço foi da ordem de:

- R: ~ 2,5 vezes menos espaço

In [58]:
df_estabelecimentos.write.parquet("dados/estabelecimentos/estabelecimentos_parquet/")

In [69]:
!du -hcs dados/estabelecimentos/*

3.7G	dados/estabelecimentos/estabelecimentos
1.5G	dados/estabelecimentos/estabelecimentos_parquet
5.2G	total


#### Pergunta 4: Vamos usar Spark SQL para obter algumas contagens. Primeiro, vamos ver quantos estabelecimentos não tem logradouro cadastrado. Use `estabelecimentos_df.createTempView("estabelecimentos")` para criar uma tabela temporária de logradouros e depois execute uma consulta SQL como `SELECT COUNT(*) FROM estabelecimentos WHERE LOGRADOURO IS NULL` para contar quantos estabelecimentos não tem logradouro informado. Quantos são?
- R: 828

In [104]:
df_estabelecimentos_parquet = spark.read.parquet("dados/estabelecimentos/estabelecimentos_parquet/")

In [ ]:
df_estabelecimentos.createTempView("estabelecimentos_csv")

In [108]:
df_estabelecimentos_parquet.createTempView("estabelecimentos_parquet")

In [109]:
spark.sql("SELECT COUNT(*) FROM estabelecimentos_parquet WHERE LOGRADOURO IS NULL")

+--------+
|count(1)|
+--------+
|     828|
+--------+

#### Pergunta 5: Em muitos casos, as UDFs (funções definidas pelo usuário) são uma forma muito conveniente de implementar uma lógica dentro de uma função python e chamá-la dentro de uma consulta SQL. Vamos fazer isso para computar quantos estabelecimento têm um logradouro  cujo endereço é uma AVENIDA. Crie uma função em Python `def is_avenida(logradouro):` que recebe um logradouro e decide se ele é uma avenida ou não. Dica: Use `startswith()` e fique atento ao resultado da questão anterior, pois ele pode impactar a lógica da sua função. Esteja atento a maiúsculas e minúsculas.Em seguida, use Spark SQL para executar uma consulta como `SELECT COUNT(*) FROM estabelecimentos WHERE is_avenida(LOGRADOURO) == True` para contar quantos estabelecimentos ficam localizados em uma avenida.
- R: 52587

In [100]:
def is_avenida(logradouro):
  if logradouro == '' or type(logradouro) == 'NoneType' or logradouro is None:
    return False
  else:
    return logradouro.lower().startswith("avenida")

logradouro1 = "Avenida 1"
logradouro2 = "rua 2"
logradouro3 = ""
print(is_avenida(logradouro1))
print(is_avenida(logradouro2))
print(is_avenida(logradouro3))

True
False
False


In [98]:
from pyspark.sql.functions import udf

isAvenidaUDF = udf(lambda z: is_avenida(z))
spark.udf.register("isAvenidaUDF", is_avenida)

<function __main__.is_avenida(logradouro)>

In [111]:
spark.sql("SELECT COUNT(*) FROM estabelecimentos_parquet WHERE isAvenidaUDF(LOGRADOURO) == True")

+--------+
|count(1)|
+--------+
|   52587|
+--------+

#### Pergunta 6: Quantos CEPs distintos existem entre os estabelecimentos?
- R: 889885


In [112]:
from pyspark.sql.functions import countDistinct

distinct_count_ceps= df_estabelecimentos_parquet.select(countDistinct("CEP"))
distinct_count_ceps.show()

+-------------------+
|count(DISTINCT CEP)|
+-------------------+
|             889885|
+-------------------+



#### Pergunta 7: Quantos CNAEs existem na tabela de CNAES?
- R: 1359

In [113]:
df_cnaes.count()

1359

#### Pergunta 8: Vários CNAEs são de cultivo, como “cultivo de milho”, “cultivo de trigo” e “cultivo de arroz” Quantos estabelecimentos possuem um CNAE relacionado a cultivo? Dicas: use a operação JOIN para criar uma dataframe que tem os estabelecimentos e as descrições de seus CNAES. Crie uma UDF que verifica se a descrição do CNAE é sobre cultivo; e usar Spark SQL para escrever uma consulta como `SELECT COUNT(*) from estabelecimentos_with_cnae WHERE is_cnae_cultivo(DESCRICAO_CNAE) == True`.
- R: 200243

In [119]:
def is_cnae_cultivo(descricao_cnae):
  if descricao_cnae == '' or type(descricao_cnae) == 'NoneType' or descricao_cnae is None:
    return False
  else:
    return descricao_cnae.lower().startswith("cultivo")

cultivo1 = "Cultivo 1"
cultivo2 = "cultivo 2"
cultivo3 = ""
print(is_cnae_cultivo(cultivo1))
print(is_cnae_cultivo(cultivo2))
print(is_cnae_cultivo(cultivo3))

True
True
False


In [115]:
df_cnaes.createTempView("cnaes")

In [120]:
isCNAEUDF = udf(lambda z: is_cnae_cultivo(z))
spark.udf.register("isCNAEUDF", is_cnae_cultivo)

<function __main__.is_cnae_cultivo(descricao_cnae)>

In [121]:
spark.sql(
"""
  SELECT
    COUNT(*)
  FROM
    estabelecimentos_parquet e
  JOIN
    cnaes c
  ON
    e.CNAE_PRINCIPAL = c.CNAE
  WHERE
     isCNAEUDF(c.DESCRICAO_CNAE) == True
"""
)

+--------+
|count(1)|
+--------+
|  200243|
+--------+

#### Pergunta 9: Quais destes provedores oferecem serviços gerenciados para uso do Apache Spark?
- R: Todas as alternativas.

#### Pergunta 10: Qual dessas características se aplica à API de Dataframes do Spark, mas não à API de RDDs?
- R: lida com dados estruturados.

#### Pergunta 11: Quantos estabelecimentos são filiais? Consulte o dicionário dos dados e use Spark SQL.
- R: 1093082

In [114]:
spark.sql(
"""
  SELECT
    COUNT(*)
  FROM
    estabelecimentos_parquet
  WHERE
    IDENTIFICADOR_MATRIZ_FILIAL == 2
"""
)

+--------+
|count(1)|
+--------+
| 1093082|
+--------+

#### Pergunta 12: Qual dessas tarefas você diria que NÃO está no escopo de engenharia de dados?

- R: Criação de modelos de inteligência artificial.

#### Pergunta 13: Caso você queira construir uma aplicação que lê mensagens do Twitter e utiliza um modelo de aprendizado de máquina para classificar o conteúdo dos tweets entre carregando sentimentos “positivos” e “negativos”, quais componentes do Spark você utilizaria?

- R: Spark Streaming e Spark ML.

#### Pergunta 14: Quais formatos o Spark consegue NÃO possuir uma API para ler de forma nativa?

- R: XML.

#### Pergunta 15: O Apache Spark é tudo isso, EXCETO…

- R: … um sistema de arquivos distribuído.